Make metatitle translatable

mantebridts opened the pull request:
Make 'page' in the metatitle translatable through subexpression in the metatitle helper

Closes #9466 and #9385

mantebridts commented:
Now I don't get it. 😕 

At first, I made a mistake and the tests failed. I then corrected the mistake, checked the tests, but they keep failing now...

But, when I checkout the master branch, the same test (about the AMP-image) keeps failing. So, can I presume I did nothing wrong? 😄

Can somebody help me get this right? 🙏

juan-g commented:
Well, let's say we want to translate page titles such as
```
My blog (Page 10)
```
into e.g. Chinese like
```
我的博客（第10页）
```
I've not tested it, however I think something you could try is having just a text placeholder like `%pagenumber%` in the file _core/server/data/meta/title.js_. For example:
```
pageString = ' (%pagenumber%)';
```
instead of
```
pageString = ' (Page ' + pagination.page + ')';
```
The `pagination.page` logic could be moved to _core/server/helpers/meta_title.js_.

So, if in the theme's _default.hbs_ template we have
```
<title>{{meta_title page=(t "Page %")}}</title>
```
instead of the current
```
<title>{{meta_title}}</title>
```
and e.g. a Chinese translation file _locales/zh.json_ including the line
```
 "Page %": "第%页",
```
then `(t "Page %")` becomes `第%页`, and this and the title from _core/server/data/meta/title.js_ both go to _core/server/helpers/meta_title.js_, where some code should be added to replace `%` with the value of  `pagination.page` (e.g. `10`) in `第%页`, and `%pagenumber%` with e.g. `第10页` in the title. And that's it.

mantebridts commented:
Thx @juan-g for the help! 🙏 

I moved the logic and tested your approach. It is possible now to use the helper like this: 
`{{meta_title page=(t " Page %")}}`

It's translatable and it is possible to place the '%' where you like in your translation (so the Chinese-speaking-Ghost-users can use it too!). 😄 

The '%' is replaced by the number of the page you're actively visiting, this logic is in _meta_title.js_
Afterwards, the title is put in _pagestring_, so it can be used like before in _title.js_. 

I think this should do the trick. 😃

mantebridts commented:
Woohooow! I finally succeeded in passing all the tests. 😅
Is there anything else I should be doing now? Let me know!

kirrg001 commented on core/server/data/meta/title.js:
```diff
@@ -12,7 +12,7 @@ function getTitle(data, root, options) {
 
     if (pagination && pagination.total > 1) {
-        pageString = ' (Page ' + pagination.page + ')';
+        pageString = pagination.pageString || ' (Page ' + pagination.page + ')';
```
The meta title is also called when using `{{ghost_head}}`. That means this solution only works when i use the `{{meta_title}}` helper.

mantebridts replied:
I don't understand. When I look to the output of <code>{{ghost_head}}</code>, I don't see any 'page' string that could be translated. 
Is it in <code>ghost_head.js</code> that the meta title is also called upon? Can you show me where, so I can investigate, cause I can't find it. 😞

mantebridts replied:
I looked into this and noticed the `{{ghost_head}}`-helper only uses `title.js` for generating the meta-tags (for twitter, facebook, etc...). These meta-tags are not printed in a paged context. See line 155 in `ghost_head.js`: 
<code>if (**!**_.includes(context, 'paged') && useStructuredData) {
                    head.push('');
                    head.push.apply(head, finaliseStructuredData(metaData));
                    head.push('');</code>

Knowing this, I think my solution doesn't form any problem. When translation is needed (in a paged context), the translated string is not printed in the meta-tags.

Unless, this printing of the meta-tags should happen in a paged context. The if-statement was added a long time ago in this commit by @cobbspur: https://github.com/TryGhost/Ghost/commit/a48458b73dc1c8bf0a730d33e3461cad512c61f6 (rewritten here: 5102637b8e23c0d33704406003a2a81d08752f37), to fix bug #5146.

Can I remove this check, so meta-tags get printed in a paged context? When I tested it here, it worked perfectly (error pages have their own template now). If this is possible, I can move the translation logic.

@kirrg001 or @cobbspur, can you tell me what to do? 😃

kirrg001 replied:
Yeah you are right. The `{{ghost_head}}` only generates the meta description. Sorry for confusion 🙈

mantebridts commented:
I restored the two files from the master-branch. That should be OK now. :)

mantebridts commented:
@kirrg001, I moved the logic and added an extra parameter, like you suggested. Works here, but the tests keep failing on some authentication logic (even on the master branch, without my changes).
Can you see if my changes are OK like this? 😃

kirrg001 commented on core/server/helpers/meta_title.js:
```diff
@@ -3,9 +3,10 @@
 // We use the name meta_title to match the helper for consistency:
 module.exports = function meta_title(options) { // eslint-disable-line camelcase
-    return getMetaDataTitle(this, options.data.root);
+    return getMetaDataTitle(this, options.data.root, options.hash);
```
Could you pls pass the whole `options` object?

Otherwise it's very hard to read where e.g. `options.page` is coming from.
Also, would you mind adding a simple unit test?

https://github.com/TryGhost/Ghost/blob/master/core/test/unit/data/meta/title_spec.js

mantebridts replied:
Will do! 👍

kirrg001 replied:
👍

kirrg001 commented:
> Works here, but the tests keep failing on some authentication logic (even on the master branch, without my changes).

Yeah, i pushed a commit in the morning. If you checkout master, pull and then rebase your branch against master, it should pass again.

kirrg001 commented:
@mantebridts Any chance you can finish this by tomorrow? Tuesday is release day. Thanks :)

mantebridts commented:
@kirrg001, I added a unit-test, hopefully everything is OK for release? 😃

mantebridts commented on core/server/data/meta/title.js:
```diff
@@ -12,7 +12,7 @@ function getTitle(data, root, options) {
 
     if (pagination && pagination.total > 1) {
-        pageString = ' (Page ' + pagination.page + ')';
+        pageString = _.has(options.hash, 'page') ? options.hash.page.replace('%', pagination.page) : ' (Page ' + pagination.page + ')';
```
Made it `pageString`. :)

kirrg001 commented on core/server/data/meta/title.js:
```diff
@@ -12,7 +12,7 @@ function getTitle(data, root, options) {
 
     if (pagination && pagination.total > 1) {
-        pageString = ' (Page ' + pagination.page + ')';
+        pageString = _.has(options.hash, 'pageString') ? options.hash.pageString.replace('%', pagination.page) : ' (Page ' + pagination.page + ')';
```
@ErisDS 

`{{meta_title pageString=(t...)}}`

Design Approval for translating the meta title if paginated.

kirrg001 replied:
@mantebridts We had an internal chat about this.

`{{meta_title page=(t...)}}`

This is the version you've used in the first commit already - so your initial idea was for now the best.

mantebridts replied:
OK, so I revert this last commit and will change it back to 'page'!

juan-g commented:
Good work. :clap:
